File: src/bot_server.cpp
#include <iostream>
#include <string>
#include <unistd.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>

class TcpServerNode : public rclcpp::Node
{
public:
    TcpServerNode()
        : Node("bot_server_node")
    {
        // Create a TCP server socket
        server_sockfd = socket(AF_INET, SOCK_STREAM, 0);
        if (server_sockfd == -1) {
            RCLCPP_ERROR(this->get_logger(), "Socket creation failed!");
            exit(1);
        }

        server_address.sin_family = AF_INET;
        server_address.sin_port = htons(2000);  // Port
        server_address.sin_addr.s_addr = INADDR_ANY; // Listen on all interfaces

        if (bind(server_sockfd, (struct sockaddr *)&server_address, sizeof(server_address)) < 0) {
            RCLCPP_ERROR(this->get_logger(), "Bind failed!");
            exit(1);
        }

        if (listen(server_sockfd, 1) < 0) {
            RCLCPP_ERROR(this->get_logger(), "Listen failed!");
            exit(1);
        }

        RCLCPP_INFO(this->get_logger(), "TCP Server Node started. Waiting for connection...");

        // Create a publisher for turtlesim movement
        pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);

        // Accept a connection from the client
        int client_sockfd = accept(server_sockfd, NULL, NULL);
        if (client_sockfd < 0) {
            RCLCPP_ERROR(this->get_logger(), "Accept failed!");
            exit(1);
        }

        RCLCPP_INFO(this->get_logger(), "Client connected. Waiting for commands...");

        // Start receiving commands from the client
        handle_client(client_sockfd);
    }

    ~TcpServerNode()
    {
        close(server_sockfd);
    }

private:
    int server_sockfd;
    struct sockaddr_in server_address;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_;

    void send_command(const std::string &command)
    {
        auto twist_msg = geometry_msgs::msg::Twist();

        // Send appropriate motion commands to the turtle
        if (command == "move_forward") {
            twist_msg.linear.x = 1.0;
        }
        else if (command == "move_backward") {
            twist_msg.linear.x = -1.0;
        }
        else if (command == "turn_left") {
            twist_msg.angular.z = 1.0;
        }
        else if (command == "turn_right") {
            twist_msg.angular.z = -1.0;
        }
        else if (command == "stop") {
            twist_msg.linear.x = 0.0;
            twist_msg.angular.z = 0.0;
        }

        pub_->publish(twist_msg);
    }

    void handle_client(int client_sockfd)
    {
        char buffer[1024];
        while (true) {
            int read_size = read(client_sockfd, buffer, sizeof(buffer));
            if (read_size == 0) {
                RCLCPP_INFO(this->get_logger(), "Client disconnected.");
                break;
            }

            buffer[read_size] = '\0';
            std::string command(buffer);

            RCLCPP_INFO(this->get_logger(), "Received command: %s", command.c_str());

            // Send the command to control the turtle
            send_command(command);
        }
        close(client_sockfd);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TcpServerNode>());
    rclcpp::shutdown();
    return 0;
}

File: tcp_examples/bot_server.py
#!/usr/bin/env python3

import socket
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import String

class TcpServerNode(Node):
    def __init__(self):
        super().__init__('bot_server_node')

        # Set up TCP server socket
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server_socket.bind(('127.0.0.1', 2000))  # Bind to localhost and port 2000
        self.server_socket.listen(1)
        self.get_logger().info("TCP server listening on port 2000")

        # Create a publisher for controlling the turtle
        self.publisher = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)

        # Accept connections from the client
        self.client_socket, self.client_address = self.server_socket.accept()
        self.get_logger().info(f"Connection from {self.client_address}")

        # Start listening to commands from the client
        self.listen_to_client()

    def listen_to_client(self):
        try:
            while True:
                # Receive data from the client
                data = self.client_socket.recv(1024)
                if not data:
                    break  # Exit if no data is received
                command = data.decode()
                self.get_logger().info(f"Received command: {command}")

                # Control the turtle based on the command
                self.control_turtle(command)

        except Exception as e:
            self.get_logger().error(f"Error while receiving data: {e}")
        finally:
            self.client_socket.close()

    def control_turtle(self, command: str):
        twist_msg = Twist()

        if command == "move_forward":
            twist_msg.linear.x = 2.0  # Move forward at speed 2
            twist_msg.angular.z = 0.0  # No rotation
        elif command == "move_backward":
            twist_msg.linear.x = -2.0  # Move backward at speed -2
            twist_msg.angular.z = 0.0  # No rotation
        elif command == "turn_left":
            twist_msg.linear.x = 0.0  # No forward movement
            twist_msg.angular.z = 1.0  # Turn left
        elif command == "turn_right":
            twist_msg.linear.x = 0.0  # No forward movement
            twist_msg.angular.z = -1.0  # Turn right
        elif command == "stop":
            twist_msg.linear.x = 0.0  # Stop moving
            twist_msg.angular.z = 0.0  # Stop rotating

        # Publish the twist message to the turtle
        self.publisher.publish(twist_msg)

def main(args=None):
    rclpy.init(args=args)
    server = TcpServerNode()
    rclpy.spin(server)
    server.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
